src/os_net/os_net.c

```diff
@@ -593,3 +678,125 @@ int OS_CloseSocket(int socket)
 #endif /* WIN32 */
 }
 
+
+/*
+ * OS_DecodeAddrinfo() will decode the contents of an addrinfo structure and
+ * log the IP version, address, and port number for each item in the
+ * linked list of addrinfo structs.
+ */
+
+int OS_DecodeAddrinfo (struct addrinfo *res) {
+    struct addrinfo *p;			/* pointer to addrinfo structs */
+
+    for (p = res; p != NULL; p = p->ai_next)
+        verbose ("%s\n",OS_DecodeSockaddr (p->ai_addr));
+    return 0;
+}
+
+
+/*
+ * OS_DecodeSockaddr() will decode a socket address and return a string with
+ * the IP version, address, and port number.
+ */
+
+char *OS_DecodeSockaddr (struct sockaddr *sa) {
+    int rc;				/* return code */
+    char ipaddr[INET6_ADDRSTRLEN];	/* printed address */
+    char ipport[NI_MAXSERV];		/* printed port */
+    static char buf[256];		/* message buffer */
+
+    rc = getnameinfo ((struct sockaddr *) sa, sa->sa_len, ipaddr,
```

Comment:
As you might have seen from Travis: `os_net/os_net.c:708:49: error: ‘struct sockaddr’ has no member named ‘sa_len’`
Changing the `sa->sa_len` to `sizeof(sa)` allows it to compile on Linux.

Reply:
The issue with Solaris is not related to this change. The code you referred to in src/os_net/os_net.c in function OS_DecodeSockaddr() is simply a function to list the contents of a sockaddr struct (passed as a pointer) for diagnostic purposes. It does not change the contents of the struct or modify it in any way.

Prior to the change to using getaddrinfo() with AF_UNSPEC, OSSEC would only operate on IPv6 addresses (IPv4 addresses were mapped onto IPv6 addresses). The modification, when combined with multiple address retrieval and select() calls, allowed all of the addresses assigned to all network interfaces on the machine to participate with OSSEC.

If you are experiencing issues with creating/accessing sockets in the queue directory, it is likely you are having permission problems with the directories. The /var/ossec/queue directory should be chmod 550 and owned by owner:group root:ossec. The ossec subdirectory under it should be chmod 750 and owned by ossec:ossec. The socket named queue that is created inside that subdirectory is chmod 660 and owned by ossec:ossec - this socket should be created automatically by OSSEC.

If you are encountering problems with permissions, try changing the directory permissions to chmod 777 on /var/ossec/queue and /var/ossec/queue/ossec . If it works, reset permissions on /var/ossec/queue and try again. If that works, try adjusting permission on /var/ossec/queue/ossec. You will be able to isolate it this way. If you have to run with relaxed permissions, you can also try using the sticky bit on the /var/ossec/queue/ossec directory - that will limit access to the creating owner and group. You can set the sticky bit on a directory using:

chmod 1777 /var/ossec/queue/ossec

Good luck!

Dave Stoddard
Network Alarm Corporation